org.eclipse.jdt.text.tests/src/org/eclipse/jdt/text/tests/folding/FoldingWithShowSelectedElementTests.java

```diff
@@ -0,0 +1,190 @@
+		JavaProjectHelper.delete(jProject);
+		JavaPlugin.getDefault().getPreferenceStore().setToDefault(PreferenceConstants.EDITOR_SHOW_SEGMENTS);
+		JavaPlugin.getDefault().getPreferenceStore().setToDefault(PreferenceConstants.EDITOR_FOLDING_CUSTOM_REGIONS_ENABLED);
+	}
+
+	@Test
+	public void testFoldingActive() throws Exception {
+		String str= """
+				package org.example.test;
+				public class A {
+					void someMethod() {
+						// this method should be folded
+					}
+				}
+				""";
+		List<IRegion> regions= FoldingTestUtils.getProjectionRangesOfFile(packageFragment, "B.java", str);
+		assertEquals(1, regions.size());
+		FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(regions, str, 2, 3);
+	}
+
+	@Test
+	public void testInsertText() throws Exception {
+		JavaPlugin.getDefault().getPreferenceStore().setValue(PreferenceConstants.EDITOR_FOLDING_CUSTOM_REGIONS_ENABLED, true);
+		String str= """
+				package org.example.test;
+				public class A {
+					// region
+					void someMethod() {
+						// content here
+					}
+					// endregion
+				}
+				""";
+
+		ICompilationUnit cu= packageFragment.createCompilationUnit("A.java", str, true, null);
+		JavaEditor editor= (JavaEditor) EditorUtility.openInEditor(cu);
+
+		try {
+			editor.setSelection(cu.getElementAt(str.indexOf("someMethod")));
+
+			ProjectionAnnotationModel model= editor.getAdapter(ProjectionAnnotationModel.class);
+
+			assertEquals("""
+						// region
+						void someMethod() {
+							// content here
+						}
+					""", editor.getViewer().getTextWidget().getText());
+
+			List<IRegion> regions= extractRegions(model);
+
+			assertEquals(2, regions.size());
+
+			IDocument document= editor.getViewer().getDocument();
+			FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(regions, str, 2, 6);
+			FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(regions, str, 3, 4);
+
+			document.replace(str.indexOf("content"), 0, "method ");
+
+			regions = extractRegions(model);
+			str = document.get();
+
+			assertEquals(2, regions.size());
+
+			FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(regions, str, 2, 6);
+			FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(regions, str, 3, 4);
+		} finally {
+			editor.close(false);
+		}
+	}
+
+	@Test
+	public void testWithTrailingWhitespaceAfterClosingBrace() throws Exception {
+		JavaPlugin.getDefault().getPreferenceStore().setValue(PreferenceConstants.EDITOR_FOLDING_CUSTOM_REGIONS_ENABLED, true);
+		String str= """
+				package org.example.test;
+				public class A {
+					void someMethod() {
+						// content here
+					}\t\t\t\t
+				}
+				""";
+
+		ICompilationUnit cu= packageFragment.createCompilationUnit("A.java", str, true, null);
+		JavaEditor editor= (JavaEditor) EditorUtility.openInEditor(cu);
+
+		try {
+			editor.setSelection(cu.getElementAt(str.indexOf("someMethod")));
+
+			assertEquals("""
+						void someMethod() {
+							// content here
+						}\t\t\t\t
```

Comment:
Including trailing whitespace seems to be necessary because otherwise it would hide the closing `}`.